### ISSUE TITLE
How do you merge multiple pdfs into a single document

### ISSUE THREAD

**Issue by lukas-loering** [ancwrd1/qpdf-rs#6]
Hey there,

first of all thank you for providing this library!

How would one go about reading multiple pdfs from memory and merging them into one document? I have looked at the test and docs you provided and the closest thing to what I need seems to be the `copy_from_foreign` function. Is this the right apporach?

I have had issues with other libraries when merging pdfs, where fonts/references to fonts where not correctly merged.

Thank you!

**Comment by ancwrd1** [comment]
Yes something like this would work, although I haven't tested it with a variety of PDFs. You could try for your particular use case.

```rust
use qpdf::QPdf;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let first = QPdf::read("/tmp/first.pdf")?;
    let second = QPdf::read("/tmp/second.pdf")?;

    for page in second.get_pages()? {
        let copied = first.copy_from_foreign(page);
        first.add_page(copied, false)?;
    }
    first.writer().write("/tmp/combined.pdf")?;

    Ok(())
}
```

**Comment by ancwrd1** [comment]
Actually foreign copy is not even needed, so it seems.

**Comment by lukas-loering** [comment]
Thank you for the reply! 

For documents with a single page this worked, however merging pdfs with multiple pages resulted in a segementation fault when accessing the merged pdf document. This is the code I tried:

```rust
fn merge_pdfs_alt<T: AsRef<[u8]>>(documents: impl IntoIterator<Item = T>) -> Result<QPdf> {
    let merged_pdf = QPdf::empty();
    trace!("initialized empty pdf");
    for (i, document) in documents.into_iter().enumerate() {
        let document = QPdf::read_from_memory(document)?;
        trace!("doc {i}: read document from buffer");
        for (j, page) in document.get_pages()?.into_iter().enumerate() {
            merged_pdf.add_page(page, false)?;
            trace!("doc {i}: added page {j}");
        }
    }
    trace!("pdf merge complete");
    Ok(merged_pdf)
}
/// The calling code
#[test]
    fn merge_pdf_test() {
        simplelog::SimpleLogger::init(LevelFilter::Trace, Default::default()).unwrap();
        let pdf1 = std::fs::read("./pdf1.pdf").unwrap();
        let pdf2 = std::fs::read("./pdf2.pdf").unwrap();

        let merged = assert_ok!(merge_pdfs_alt([pdf1, pdf2]));
        merged
            .writer()
            .write("./merged.pdf")
            .unwrap();
    }
}
```
The following is the log output:
```
10:18:57 [TRACE] (2) r2d2_pdf_forms: [src/lib.rs:126] initialized empty pdf
10:18:57 [TRACE] (2) r2d2_pdf_forms: [src/lib.rs:129] doc 0: read document from buffer
10:18:57 [TRACE] (2) r2d2_pdf_forms: [src/lib.rs:132] doc 0: added page 0
10:18:57 [TRACE] (2) r2d2_pdf_forms: [src/lib.rs:132] doc 0: added page 1
10:18:57 [TRACE] (2) r2d2_pdf_forms: [src/lib.rs:132] doc 0: added page 2
10:18:57 [TRACE] (2) r2d2_pdf_forms: [src/lib.rs:129] doc 1: read document from buffer
10:18:57 [TRACE] (2) r2d2_pdf_forms: [src/lib.rs:132] doc 1: added page 0
10:18:57 [TRACE] (2) r2d2_pdf_forms: [src/lib.rs:132] doc 1: added page 1
10:18:57 [TRACE] (2) r2d2_pdf_forms: [src/lib.rs:132] doc 1: added page 2
10:18:57 [TRACE] (2) r2d2_pdf_forms: [src/lib.rs:135] pdf merge complete
error: test failed, to rerun pass `-p r2d2_pdf_forms --lib`

Caused by:
  process didn't exit successfully: `/.../pdfutils/target/debug/deps/r2d2_pdf_forms-275c9e1e3a367dec 'tests::merge_pdf_test' --format=json --exact -Z unstable-options --show-output` (signal: 11, SIGSEGV: invalid memory reference)
```

Any ideas on this? Is my usage incorrect or is this a bug? 
Again, thanks for the help. I also understand if you dont have the time/dont want to play support for users, just let me know!

**Comment by ancwrd1** [comment]
Could you try with copy_from_foreign approach above?

**Comment by lukas-loering** [comment]
I have changed the inner loop to this:
```rust
  for (j, page) in document.get_pages()?.into_iter().enumerate() {
      let copied = merged_pdf.copy_from_foreign(page);
      merged_pdf.add_page(copied, false)?;
      trace!("doc {i}: added page {j}");
  }
```
This solves this segementation fault but sadly produces an unsable pdf. The following should read "ABC" but this is written instead:
![grafik](https://github.com/ancwrd1/qpdf-rs/assets/52287649/e26ae6b9-3203-46d4-99a1-5b7e9821cfc5)

My first approach, without `copy_from_foreign` did not have this issue.

**Comment by ancwrd1** [comment]
Interesting. Could be an upstream issue with qpdf (both of them).
I think I should try perhaps a newer version, there were many updates already.

**Comment by ancwrd1** [comment]
So the crash issue is most like related to the QPdf going out of scope (on each loop iteration) so the memory is released and the page reference becomes invalid. You could try constructing an array of QPdf objects first and then a merged document from each array element.
I think I can fix this one.

**Comment by lukas-loering** [comment]
🥳 I was able to fix it!

I suspect you were correct that the QPdf was dropped after each iteration. I loaded the documents into a vector and iterated over that. Also, I made sure to write the pdf to memory instead of returning the merged QPpdf, as the `add_page()` function ony seemed to add a reference to the page. As the documents array was dropped at the end of the function, those became invalid.

Thank you very much for your help!

This is the code which now works:
```rust
fn merge_pdfs_alt<T: AsRef<[u8]>>(documents: impl IntoIterator<Item = T>) -> Result<Vec<u8>> {
    let merged_pdf = QPdf::empty();
    trace!("initialized empty pdf");
    let documents = documents
        .into_iter()
        .map(QPdf::read_from_memory)
        .collect::<std::result::Result<Vec<QPdf>, QPdfError>>()?;
    trace!("read {} documents from memory", documents.len());
    for (i, document) in documents.iter().enumerate() {
        for (j, page) in document.get_pages()?.iter().enumerate() {
            // let copied = merged_pdf.copy_from_foreign(page);
            merged_pdf.add_page(page, false)?;
            trace!("doc {i}: added page {j}");
        }
    }
    trace!("pdf merge complete");
    let buffer = merged_pdf.writer().write_to_memory()?;
    trace!("written to memory");
    Ok(buffer)
}
```

**Comment by ancwrd1** [comment]
Hmm, that actually should not happen, crashes are not expected in the Rust programs. to be honest. Are you using the latest qpdf crate version? There was a fix added recently for the memory-loaded PDFs.
I also can't reproduce it here.

**Comment by lukas-loering** [comment]
I am using qpdf version 1.6.0. If you want to investigate this further I can try to create a minimum example for you. Unfortunately I cannot share the actual pdfs that crashed as they contain sensitive information.

**Comment by ancwrd1** [comment]
Should be  fixed in commit 9383b6b13995e5e6bc4c4a83b583383f5936a850 and version 0.1.7.
No need to keep the documents in the scope anymore.
